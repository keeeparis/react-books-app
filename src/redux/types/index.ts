export interface mainState {
  startIndex: number
  skip: boolean
  input: string
  totalItems: number
  books: Book[]
  category: Category
  sorting: Sorting
}

export interface Book {
  etag: string
  id: string
  volumeInfo: volumeInfo
}

export interface volumeInfo {
  authors?: string[]
  imageLinks?: {
    smallThumbnail: string
    thumbnail: string
  }
  categories?: string[]
  pageCount?: number
  description?: string
  language?: string
  title?: string
  publishedDate?: string
}

export enum Category {
  ALL = 'All',
  ART = 'Art',
  BIOGRAPHY = 'Biography',
  COMPUTERS = 'Computers',
  HISTORY = 'History',
  MEDICAL = 'Medical',
  POETRY = 'Poetry',
  FICTION = 'Fiction',
}

export enum Sorting {
  RELEVANCE = 'relevance',
  NEWEST = 'newest',
}
