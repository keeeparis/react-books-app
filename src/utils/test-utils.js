import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import 'regenerator-runtime/runtime'
import { Provider } from 'react-redux'
import { store } from '../redux/store/store'
import { MemoryRouter } from 'react-router-dom'
import '../i18n/config'

const AllTheProviders = ({ children }) => {
  return (
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  )
}

const customRender = (ui, options) => {
  return render(ui, { wrapper: AllTheProviders, ...options })
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { userEvent }
export * from 'regenerator-runtime/runtime'
export * from '../i18n/config'

export { customRender as render }
