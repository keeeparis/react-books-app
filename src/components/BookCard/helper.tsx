import { volumeInfo } from '../../redux/types'
import styles from './BookCard.module.scss'

export const transformCategory = (categories: volumeInfo['categories']) => {
  const categoryDiv = categories ? categories[0] : <>&nbsp;</>
  return categoryDiv
}

export const transformImage = (images: volumeInfo['imageLinks']) => {
  const imagesDiv = images ? (
    <img src={images.thumbnail} alt="picture" loading="lazy" />
  ) : (
    <div>Нет изображения</div>
  )
  return imagesDiv
}

export const transformAuthors = (authors: volumeInfo['authors']) => {
  if (!authors) return
  const authorsDiv = authors.map((author, i) => (
    <div className={styles.author} key={author}>
      {author}
      {i + 1 !== authors.length ? ',' : null}
    </div>
  ))
  return authorsDiv
}
