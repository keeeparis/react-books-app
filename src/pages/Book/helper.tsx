import { volumeInfo } from '../../redux/types'

export const transformCategory = (categories: volumeInfo['categories']) => {
  const categoryDiv =
    categories && categories.map((cat) => <div key={cat}>{cat}</div>)
  return categoryDiv
}

export const transformAuthors = (authors: volumeInfo['authors']) => {
  if (!authors) return
  const authorsDiv = authors.map((author) => <h4 key={author}>{author}</h4>)
  return authorsDiv
}
