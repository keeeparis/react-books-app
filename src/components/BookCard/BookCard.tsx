import React from 'react'
import { Link } from 'react-router-dom'
import { CSSTransition } from 'react-transition-group'
import { volumeInfo } from '../../redux/types'
import styles from './BookCard.module.scss'
import { transformAuthors, transformCategory, transformImage } from './helper'

interface BookCardProps {
  book: volumeInfo
  id: string
}

const BookCard = React.memo(({ book, id }: BookCardProps) => {
  const category = transformCategory(book.categories)
  const image = transformImage(book.imageLinks)
  const authors = transformAuthors(book.authors)

  return (
    <Link to={`/book/${id}`}>
      <CSSTransition
        classNames="fade"
        addEndListener={() => {}}
        timeout={300}
        appear={true}
        in={true}
      >
        <div className={styles.container} title={book.title}>
          <div className={styles.imageWrapper}>{image}</div>
          <div className={styles.category}>{category}</div>
          <div className={styles.title}>{book.title}</div>
          <div className={styles.authorWrapper}>{authors}</div>
        </div>
      </CSSTransition>
    </Link>
  )
})

export default BookCard
