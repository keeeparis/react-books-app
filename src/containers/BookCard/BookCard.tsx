import React from 'react'
import { Link } from 'react-router-dom'
import { volumeInfo } from '../../redux/types'
import styles from './BookCard.module.scss'
import { transformAuthors, transformCategory, transformImage } from './helper'

interface BookCardProps {
  book: volumeInfo
  id: string
}

const BookCard = React.memo(({ book, id }: BookCardProps) => {
  const category = transformCategory(book.categories)
  const image = transformImage(book.imageLinks)
  const authors = transformAuthors(book.authors)

  return (
    <Link to={`/book/${id}`}>
      <div className={styles.container}>
        <div className={styles.imageWrapper}>{image}</div>
        <div className={styles.category}>{category}</div>
        <div className={styles.title}>{book.title}</div>
        <div className={styles.authorWrapper}>{authors}</div>
      </div>
    </Link>
  )
})

export default BookCard
