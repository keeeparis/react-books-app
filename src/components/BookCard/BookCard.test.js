import BookCard from './BookCard'
import { Category } from '../../redux/types'
import { transformAuthors, transformCategory, transformImage } from './helper'
import { render, screen, fireEvent, userEvent } from 'test-utils'

const book = {
  categories: [Category.ART, Category.FICTION],
  authors: ['Conan Doyle', 'Richard Dawkins'],
  title: 'Crime and Punishment',
  imageLinks: {
    thumbnail: 'https://some-url.com',
  },
  description: 'What a beautiful story to tell about Rodion Raskolnikov.',
}

jest.mock('react-transition-group', () => {
  const FakeTransition = jest.fn(({ children }) => children)
  return { CSSTransition: FakeTransition }
})

test('test BookCard Component', () => {
  const { asFragment } = render(<BookCard book={book} id={1} />)

  expect(asFragment()).toMatchSnapshot()
})

describe('testing transform Categories', () => {
  const data = [
    { value: [], toEqual: <>&nbsp;</> },
    { value: 1, toEqual: <>&nbsp;</> },
    { value: [Category.ART], toEqual: Category.ART },
    { value: [Category.ART, Category.FICTION], toEqual: Category.ART },
  ]

  data.forEach((element) => {
    it(`testing ${element.value} to equal ${element.toEqual}`, () => {
      const expected = transformCategory(element.value)
      expect(expected).toStrictEqual(element.toEqual)
    })
  })
})

describe('testing transform Images', () => {
  const data = [
    { value: {}, toEqual: <div>Нет изображения</div> },
    { value: [], toEqual: <div>Нет изображения</div> },
    { value: 1, toEqual: <div>Нет изображения</div> },
    { value: '1', toEqual: <div>Нет изображения</div> },
    {
      value: { thumbnail: '' },
      toEqual: <div>Нет изображения</div>,
    },
    {
      value: { thumbnail: 'https://some-url.com' },
      toEqual: (
        <img src={'https://some-url.com'} alt="picture" loading="lazy" />
      ),
    },
  ]

  data.forEach((element) => {
    it(`testing ${element.value} to equal ${element.toEqual}`, () => {
      const expected = transformImage(element.value)
      expect(expected).toStrictEqual(element.toEqual)
    })
  })
})

describe('testing transform Authors', () => {
  const data = [
    { value: {}, toEqual: null },
    { value: [], toEqual: null },
    { value: 1, toEqual: null },
    { value: '1', toEqual: null },
    // {
    //   value: ['Conan Doyle'],
    //   toEqual: [
    //     <h4 className="author" key={'Conan Doyle'}>
    //       Conan Doyle
    //     </h4>,
    //   ],
    // },
    // {
    //   value: ['Conan Doyle', 'Richard Dawkins'],
    //   toEqual: [
    //     <h4 className="author">
    //       {'Conan Doyle'}
    //       {','}
    //     </h4>,
    //     <h4 className="author">Richard Dawkins</h4>,
    //   ],
    // },
  ]

  data.forEach((element) => {
    it(`testing ${element.value} to equal ${element.toEqual}`, () => {
      const expected = transformAuthors(element.value)
      expect(expected).toStrictEqual(element.toEqual)
    })
  })
})
