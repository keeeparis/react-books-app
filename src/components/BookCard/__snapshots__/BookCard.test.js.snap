// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BookCard Component 1`] = `
<DocumentFragment>
  <a
    href="/book/1"
  >
    <div
      class="container"
      title="Crime and Punishment"
    >
      <div
        class="imageWrapper"
      >
        <img
          alt="picture"
          loading="lazy"
          src="https://some-url.com"
        />
      </div>
      <div
        class="category"
      >
        art
      </div>
      <div
        class="title"
      >
        Crime and Punishment
      </div>
      <div
        class="authorWrapper"
      >
        <h4
          class="author"
        >
          Conan Doyle,
        </h4>
        <h4
          class="author"
        >
          Richard Dawkins
        </h4>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
